### PR TITLE
main : don't print special tokens with --grammar

### PR DESCRIPTION
**Pull request by jart**
The CLI interface was recently changed to print special control tokens like the </s> stop message one. This token shouldn't be printed if the grammar flag was passed, unless the grammar specifies it, because that breaks shell-scriptability.

### CODE REVIEW

**Comment by jart** [comment]
See:

- https://justine.lol/oneliners/#filename
- https://justine.lol/matmul/#:~:text=%2D%2Dgrammar%20%27root%20%3A%3A%3D%20%22yes%22%20%7C%20%22no%22%27

I like to ask LLMs yes/no questions in shell scripts. I use the grammar flag to force it to only print yes or no. If it instead prints "no<s>" or "no<|end-of-turn|>" then that breaks my shell script if statments.

**Comment by HanClinto** [comment]
That makes sense. 

I don't use shell scripts with grammars, but I wonder if this functionality would be better added behind a command line option to specifically render special tokens or hide them?  If I'm trying to debug a grammar- constrained generation, it would tend to want to display the special tokens rather than hide them. 

How do you feel about separating this flag out into its own option?

**Review by mofosyne — APPROVED** [review]
Logic checks out. Intent is sensible

What's your thought about @HanClinto  idea of separating it out into it's own flag? Regardless, should be safe to make a separate PR and merge this in anyway as the default behavior of omitting special token with grammar make sense.

---

FYI: Android CI is bit broken in main branch, but I see a PR coming in soon to fix. But this doesn't touch android anyway.

**Comment by ggerganov** [comment]
> I don't use shell scripts with grammars, but I wonder if this functionality would be better added behind a command line option to specifically render special tokens or hide them? If I'm trying to debug a grammar- constrained generation, it would tend to want to display the special tokens rather than hide them.

> How do you feel about separating this flag out into its own option?

I agree it's better to have this as separate flag - should consolidate with the existing `conversation` flag - rename it and reuse it.

**Comment by mofosyne** [comment]
Possible approach that @jart could potentially use in https://github.com/jart/llama.cpp/pull/3 to address @HanClinto 's idea of separating out the flag. Investigated @ggerganov idea of consolidating with existing conversation flag but there is significant enough difference in semantic that i could not merge it.

Feel free to adjust as needed or ignore if it puts too much complexity to this PR.

**Comment by jart** [comment]
One thing you could do is print the special tokens out of band to file descriptor 3. Then if a shell script doesn't want them, it could either pass a flag to disable them, or simply say `3>/dev/null`.

**Comment by teleprint-me** [comment]
I haven't had a chance to look at the grammar yet, but I'm wondering what the logic is behind including the special tokens in the output?

**Comment by mofosyne** [comment]
To test this you can try these

```
cmake -B build -DCMAKE_BUILD_TYPE=Debug
cmake --build build

echo "== Expect Control Token To Shared Console 3>&1 =="
./build/bin/main --hf-repo TheBloke/phi-2-GGUF --hf-file phi-2.Q6_K.gguf --grammar 'root ::= "yes" | "no"' --temp 0 -c 0 --no-display-prompt --log-disable -p "<|user|>
Say yes
<|assistant|>" 2>/dev/null 3>&1
echo
echo "== Expect No Control Token To Console because 3>/dev/null =="
./build/bin/main --hf-repo TheBloke/phi-2-GGUF --hf-file phi-2.Q6_K.gguf --grammar 'root ::= "yes" | "no"' --temp 0 -c 0 --no-display-prompt --log-disable -p "<|user|>
Say yes
<|assistant|>" 2>/dev/null 3>/dev/null
echo
echo "== Expect No Control Token To Console because 3>&- =="
./build/bin/main --hf-repo TheBloke/phi-2-GGUF --hf-file phi-2.Q6_K.gguf --grammar 'root ::= "yes" | "no"' --temp 0 -c 0 --no-display-prompt --log-disable -p "<|user|>
Say yes
<|assistant|>" 2>/dev/null 3>&-
echo
echo == Expect No Control Token To Console as we are still in grammar mode ==
./build/bin/main --hf-repo TheBloke/phi-2-GGUF --hf-file phi-2.Q6_K.gguf --grammar 'root ::= "yes" | "no"' --temp 0 -c 0 --no-display-prompt --log-disable -p "<|user|>
Say yes
<|assistant|>" 2>/dev/null
echo
echo == Expect Control Token To Console as we are in normal completion mode ==
./build/bin/main --hf-repo TheBloke/phi-2-GGUF --hf-file phi-2.Q6_K.gguf --temp 0 -c 0 --no-display-prompt --log-disable -p "<|user|>
Hi
<|assistant|>" 2>/dev/null 3>&1
echo
```

**Comment by mofosyne** [comment]
> I haven't had a chance to look at the grammar yet, but I'm wondering what the logic is behind including the special tokens in the output?

Might be handy for debugging at least.

Also it has tokens showing the split between the user, assistant and end of text. Might be handy for integration if not using the apis or libraries for some reason.

It does get me thinking if it's possible to also seperate the input special tokens so that the special tokens are fully out of band. Might help make it a little bit more secure?

File: llama.h
Comment (ggerganov) [inline llama.h:821]: No need for the `_token` suffix in the function name. Also, the API already tends to use the term "special" over "control" so better to call it:

```suggestion
    LLAMA_API bool llama_token_is_special(const struct llama_model * model, llama_token token);
```

Comment (mofosyne) [reply llama.h]: Okay, pushed a commit that drops the _token suffix

File: examples/main/main.cpp
Comment (ggerganov) [inline examples/main/main.cpp:131]: I don't think this is needed. If one wants to debug special/control tokens, then they don't pass `--no-special`. For script usage where special tokens are not wanted, you pass `--no-special`. The extra file descriptor here is too complicated

Comment (mofosyne) [reply examples/main/main.cpp]: Jart has merged in the adjustment to this PR now.

Will wait a bit further for any more feedback, but I think we have now reached a consensus point for this PR and will merge this in a day or two if no other problems emerges (Or immediate upon positive confirmation from you).

**Review by ngxson — APPROVED** [review]

**Comment by mofosyne** [comment]
@ngxson thanks. Confident now that full consensus has been reached now. Merging

**Pull request merged**
